### PR TITLE
Uses ICU4J for charset detection

### PR DESCRIPTION
**Pull request by jnioche**
maybe make this optional based on config?

### CODE REVIEW

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java [hunk 1]
```diff
@@ -237,4 +260,21 @@ public static void print(Node node, String indent, StringBuffer sb) {
...
         }
     }
+
+    /**
+     * Parse out a charset from a content type header.
+     * 
+     * @param contentType e.g. "text/html; charset=EUC-JP"
+     * @return "EUC-JP", or null if not found. Charset is trimmed and uppercased.
+     */
+    private final static String getCharsetFromContentType(String contentType) {
```
Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: You didn't want to use org.apache.http.entity.ContentType.parse

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java [hunk 2]
```diff
@@ -119,9 +129,22 @@ public void execute(Tuple tuple) {
...
         Set<String> slinks = Collections.emptySet();
 
+        String charset = null;
+
+        // check if the server specified a charset
+        String contentType = KeyValues.getValue(HttpHeaders.CONTENT_TYPE, metadata);
+        charset = getCharsetFromContentType(contentType);
+
+        // filter HTML tags
+        detector.enableInputFilter(true);
```
Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: Probably worth putting in the prepare method

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java [hunk 3]
```diff
@@ -119,9 +129,22 @@ public void execute(Tuple tuple) {
...
+        // give it a hint
+        detector.setDeclaredEncoding(charset);
+        detector.setText(content);
+        charset = detector.detect().getName();
```
Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: detector.detect() can return null

**Comment by GuiForget** [comment]
I'm ok with NOT making it optional

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java (hunk 1 above)
Comment (jnioche) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: good idea!

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java (hunk 2 above)
Comment (jnioche) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: The Detector is not serializable with Storm in deployed mode, had to move it from a field to a variable within the execute method anyway.

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java (hunk 3 above)
Comment (jnioche) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: good catch, thanks

**Comment by jnioche** [comment]
Have added your suggestions, thanks for reviewing this.

**Comment by jnioche** [comment]
Note to self : will need to add the ICUJ license explicitely

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java [hunk 4]
```diff
@@ -56,6 +62,9 @@
     private boolean ignoreOutsideHost = false;
     private boolean ignoreOutsideDomain = false;
 
+    private static final Pattern charsetPattern = Pattern
+            .compile("(?i)\\bcharset=\\s*\"?([^\\s;\"]*)");
```
Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: Not used anymore

File: jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java [hunk 5]
```diff
@@ -119,9 +127,35 @@ public void execute(Tuple tuple) {
...
+
+        // filter HTML tags
+        CharsetDetector detector = new CharsetDetector();
+        detector.enableInputFilter(true);
+        // give it a hint
+        detector.setDeclaredEncoding(charset);
+        detector.setText(content);
+        try {
+            CharsetMatch charsetMatch = detector.detect();
+            if (charsetMatch != null)
```
Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: Code style: curly braces are required

Comment (jnioche) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: shoudn't that be enforced by your formatting profile? I applied it but still had to do it by hand

Comment (GuiForget) [inline jsoup-parser/src/main/java/com/shopstyle/crawler/jsoup/JSoupParserBolt.java]: Sadly it isn't. We have it automated via what's called a Save Action, not the Formatter

**Comment by GuiForget** [comment]
I'll commit it for you so I can squash your commits

**Comment by jnioche** [comment]
great, thanks

**Pull request closed without merging**